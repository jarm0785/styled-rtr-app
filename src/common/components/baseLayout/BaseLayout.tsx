import React, { FC, ReactElement } from 'react';
import { Header, Navbar, Main, Sidebar, Footer } from '../layout/Grid';
import { Title } from '../typography/Typography';
import styles from './baseLayout.module.scss';

export interface IBaseLayoutProps {
  main: string | ReactElement;
  sidebar: string | ReactElement;
  header: string | ReactElement;
  nav: string | ReactElement;
  footer: string | ReactElement;
}

const BaseLayout: FC<Partial<IBaseLayoutProps>> = ({ header, nav, main, sidebar, footer }) => {

  return (
    // Styling thru scss modules also supported
    <div className={ styles.pageWrap }>
      <Header gridColumn="1 / -1" data-testid="base_layout:header">
        { header ?? <Title>{ 'Header' }</Title> }
      </Header>
      <Navbar gridColumn="2 / 3" data-testid="base_layout:navbar">
        { nav ?? <Title>{ 'Navbar Items' }</Title> }
      </Navbar>
      <Main gridColumn="2 / 3" data-testid="base_layout:main">
        { main ?? <Title>{ 'Main' }</Title> }
      </Main>
      <Sidebar gridColumn="1 / 2" gridRow="2 / 4" data-testid="base_layout:sidebar">
        { sidebar ?? <Title>{ 'Side Panel' }</Title> }
      </Sidebar>
      <Footer gridColumn="1 / -1" data-testid="base_layout:footer">
        { footer ?? <Title>{ 'Footer' }</Title> }
      </Footer>
    </div>
  );
};

export default BaseLayout;
