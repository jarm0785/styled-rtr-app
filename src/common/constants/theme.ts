import type { ThemeType } from '../interface/ITheme';

type ThemeConstant = Record<string, ThemeType>;

export const THEME: ThemeConstant = {
  DEFAULT: 'default',
  DARK: 'dark',
};
