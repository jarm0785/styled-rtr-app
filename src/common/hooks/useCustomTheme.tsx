import { useSelector } from 'react-redux';
import { ICustomTheme } from '../interface/ITheme';
import type { ThemeType } from '../interface/ITheme';
import globalThemeSelectors from '../store/selectors/globalThemeSelectors';


export default function useCustomTheme(): { theme: Record<ThemeType, ICustomTheme>, activeTheme: ThemeType } {
  const theme = useSelector(globalThemeSelectors.getTheme);
  const activeTheme = useSelector(globalThemeSelectors.getActiveTheme);

  return { theme, activeTheme };
}
