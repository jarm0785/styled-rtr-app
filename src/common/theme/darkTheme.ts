import { ICustomTheme } from '../interface/ITheme';

export const darkTheme: ICustomTheme = {
  colors: {
    primary: '#FFF',
    secondary: '',
    primaryBackground: 'linear-gradient(to right, #000000, #434343)',
    secondaryBackground: '',
    textPrimary: '#FFFFFF',
    textSecondary: '#FAFAFA',
    border: '#1da1f2',
  },
  button: {
    primary: {
      color: '#fff',
      borderColor: '#1da1f2',
      backgroundColor: '#1da1f2',
      hoverBackgroundColor: '#1991da'
    },
    secondary: {
      color: '#1da1f2',
      borderColor: '#1da1f2',
      backgroundColor: '#15202b',
      hoverBackgroundColor: '#162e42'
    }
  },
};
