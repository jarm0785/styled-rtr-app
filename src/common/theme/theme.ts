import { ICustomTheme } from '../interface/ITheme';

export const defaultTheme: ICustomTheme = {
  colors: {
    primary: '#14171a',
    secondary: '#657786',
    primaryBackground: 'linear-gradient(to right, #E3F2FD, #FFF);',
    secondaryBackground: '',
    textPrimary: '#14171a',
    textSecondary: '#657786',
  },
  button: {
    primary: {
      color: '#fff',
      borderColor: '#1da1f2',
      backgroundColor: '#1da1f2',
      hoverBackgroundColor: '#1991da'
    },
    secondary: {
      color: '#1da1f2',
      borderColor: '#1da1f2',
      backgroundColor: '#fff',
      hoverBackgroundColor: '#ebf5fd'
    }
  },
};
