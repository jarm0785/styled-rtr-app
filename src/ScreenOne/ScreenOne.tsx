import React, { FunctionComponent } from 'react';
import { useHistory } from 'react-router';
import BaseLayout from '../common/components/baseLayout/BaseLayout';
import { SimpleButton } from '../common/components/button/Button';
import Box from '../common/components/layout/Box';
import Flex from '../common/components/layout/Flex';
import ThemeToggle from '../common/components/themeToggle/ThemeToggle';
import { Text } from '../common/components/typography/Typography';
import { ROUTES } from '../common/constants/routes';
import ChuckNorris from './components/ChuckNorris';

const ScreenOne: FunctionComponent = () => {
  const history = useHistory();

  return (
    <BaseLayout
      main={
        <Flex justifyContent="center" alignItems="center" height="100%">
          <ChuckNorris />
        </Flex>
      }
      header={ (
        <Flex
          justifyContent="flex-end"
          alignItems="center"
        >
          <Box mr={ 1 }>
            <Text color="textPrimary" fontSize="1rem">
              Theme Toggle
            </Text>
          </Box>
          <Box>
            <ThemeToggle />
          </Box>
        </Flex>
      ) }
      nav={ (
        <Flex justifyContent="flex-end">
          <SimpleButton onClick={ () => history.push(ROUTES.SECOND_SCREEN) }>
            <Text color="textPrimary" fontSize="1rem">
              { 'Next Page >' }
            </Text>
          </SimpleButton>
        </Flex>
      )}
    />
  );
};

export default ScreenOne;
