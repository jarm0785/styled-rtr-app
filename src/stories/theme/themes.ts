import { defaultTheme } from '../../common/theme/theme';
import { darkTheme } from '../../common/theme/darkTheme';
import { ICustomTheme, ThemeType } from '../../common/interface/ITheme';

const themes: { customThemes: Record<ThemeType, { theme: ICustomTheme }> } = {
  customThemes: {
    default: {
      theme: defaultTheme,
    },
    dark: {
      theme: darkTheme,
    },
  }
};

export default themes;
