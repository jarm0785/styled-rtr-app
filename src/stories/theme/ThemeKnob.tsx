import React, { PropsWithChildren } from 'react';
import { ThemeProvider } from 'styled-components';
import { THEME } from '../../common/constants/theme';
import type { ThemeType } from '../../common/interface/ITheme';
import themes from './themes';

interface IThemeKnobProps {
  theme: ThemeType;
}

export default ({ children, theme }: PropsWithChildren<IThemeKnobProps>) => (
  <ThemeProvider theme={ themes.customThemes[theme || THEME.DEFAULT] }>
    { children }
  </ThemeProvider>
);
