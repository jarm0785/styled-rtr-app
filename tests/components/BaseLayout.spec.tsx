/**
 * @jest-environment jsdom
 */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { fireEvent, queryByTestId } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import useCustomTheme from '../../src/common/hooks/useCustomTheme';
import { render } from '../utils/test-utils';
import ScreenOne from '../../src/ScreenOne/ScreenOne';
import BaseLayout from '../../src/common/components/baseLayout/BaseLayout';
import { defaultTheme } from '../../src/common/theme/theme';

describe('Base Layout display and theme toggle', () => {
  const ThemedBaseLayout = () =>
    <ThemeProvider theme={ defaultTheme }>
      <BaseLayout />
    </ThemeProvider>;

  const ThemedScreen = () => {
    const { theme, activeTheme } = useCustomTheme();

    return (
      <ThemeProvider theme={ theme[activeTheme]}>
        <ScreenOne />
      </ThemeProvider>
    );
  };

  // Sample snapshot test with styled components.
  it('should render base layout component properly', () => {
    const componentRenderer = renderer.create(<ThemedBaseLayout/>);
    const tree = componentRenderer.toJSON();

    expect(tree).toMatchSnapshot();
  });

  // Sample Component test using providers and jsdom
  it('Should toggle between light and dark theme.', () => {
    const { container } = render(<ThemedScreen/>);
    const header = queryByTestId(container, 'base_layout:header');
    const main = queryByTestId(container, 'base_layout:header');
    const footer = queryByTestId(container, 'base_layout:header');
    const toggle = queryByTestId(container, 'switch:toggle_wrapper');
    const layoutElements = [header, main, footer];

    fireEvent.click(toggle);
    layoutElements.forEach((el) => {
      expect(el).toHaveStyle('background: linear-gradient(to right,#000000,#434343)');
    });
    fireEvent.click(toggle);
    layoutElements.forEach((el) => {
      expect(el).toHaveStyle('background: linear-gradient(to right,#E3F2FD,#FFF)');
    });
  });
});
