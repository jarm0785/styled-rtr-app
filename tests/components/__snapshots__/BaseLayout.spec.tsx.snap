// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Layout display and theme toggle should render base layout component properly 1`] = `
.c0 {
  grid-column: 1 / -1;
  background: linear-gradient(to right,#E3F2FD,#FFF);
}

.c2 {
  grid-column: 2 / 3;
  background: linear-gradient(to right,#E3F2FD,#FFF);
}

.c3 {
  grid-column: 2 / 3;
  background: linear-gradient(to right,#E3F2FD,#FFF);
}

.c4 {
  grid-column: 1 / 2;
  grid-row: 2 / 4;
  background: linear-gradient(to right,#E3F2FD,#FFF);
}

.c5 {
  grid-column: 1 / -1;
  background: linear-gradient(to right,#E3F2FD,#FFF);
}

.c1 {
  font-size: 1.5rem;
  color: #14171a;
  margin: '1rem 0';
}

<div
  className="pageWrap"
>
  <header
    className="c0"
    data-testid="base_layout:header"
  >
    <h1
      className="c1"
    >
      Header
    </h1>
  </header>
  <nav
    className="c2"
    data-testid="base_layout:navbar"
  >
    <h1
      className="c1"
    >
      Navbar Items
    </h1>
  </nav>
  <main
    className="c3"
    data-testid="base_layout:main"
  >
    <h1
      className="c1"
    >
      Main
    </h1>
  </main>
  <aside
    className="c4"
    data-testid="base_layout:sidebar"
  >
    <h1
      className="c1"
    >
      Side Panel
    </h1>
  </aside>
  <aside
    className="c5"
    data-testid="base_layout:footer"
  >
    <h1
      className="c1"
    >
      Footer
    </h1>
  </aside>
</div>
`;
